File: frontend/src/api/rag.ts
import { apiRequest } from "./client";

export type KnowledgeBaseStatus =
  "archived" | "draft" | "failed" | "indexing" | "ready";
export type RAGIndexBuildStatus =
  "cancelled" | "failed" | "queued" | "running" | "succeeded";

export interface KnowledgeBaseDatasetVersion {
  readonly dataset_version_id: string;
  readonly attached_at: string;
}

export interface KnowledgeBaseSummary {
  readonly knowledge_base_id: string;
  readonly name: string;
  readonly description: string | null;
  readonly status: KnowledgeBaseStatus;
  readonly embedding_provider: string;
  readonly embedding_model: string;
  readonly embedding_dimension: number;
  readonly attached_dataset_version_count: number;
  readonly created_at: string;
  readonly updated_at: string;
}

export interface KnowledgeBaseDetail extends KnowledgeBaseSummary {
  readonly chunking_configuration: Readonly<Record<string, unknown>>;
  readonly dataset_versions: readonly KnowledgeBaseDatasetVersion[];
  readonly active_index_build_id: string | null;
  readonly indexed_document_count: number;
  readonly indexed_chunk_count: number;
  readonly error_code: string | null;
  readonly safe_error_message: string | null;
  readonly archived_at: string | null;
}

export interface RAGIndexBuild {
  readonly index_build_id: string;
  readonly knowledge_base_id: string;
  readonly status: RAGIndexBuildStatus;
  readonly indexed_document_count: number;
  readonly indexed_chunk_count: number;
  readonly embedding_count: number;
  readonly created_at: string;
  readonly started_at: string | null;
  readonly finished_at: string | null;
  readonly cancelled_at: string | null;
  readonly error_code: string | null;
  readonly safe_error_message: string | null;
}

export interface KnowledgeBasePage {
  readonly items: readonly KnowledgeBaseSummary[];
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
}

export interface RAGIndexBuildPage {
  readonly items: readonly RAGIndexBuild[];
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
}

export interface RAGSearchResult {
  readonly chunk_id: string;
  readonly document_id: string;
  readonly dataset_version_id: string;
  readonly rank: number;
  readonly score: number;
  readonly excerpt: string;
  readonly page_number: number | null;
  readonly section: string | null;
  readonly document_title: string;
}

export interface RAGSearchResponse {
  readonly knowledge_base_id: string;
  readonly results: readonly RAGSearchResult[];
  readonly insufficient_evidence: boolean;
}

function queryString(
  values: Readonly<Record<string, number | string | undefined>>,
): string {
  const query = new URLSearchParams();
  for (const [key, value] of Object.entries(values)) {
    if (value !== undefined && value !== "") query.set(key, String(value));
  }
  return `?${query.toString()}`;
}

export function listKnowledgeBases(
  options: {
    readonly limit?: number;
    readonly offset?: number;
    readonly signal?: AbortSignal;
    readonly status?: KnowledgeBaseStatus;
  } = {},
): Promise<KnowledgeBasePage> {
  return apiRequest(
    `/ai/rag/knowledge-bases${queryString({ limit: options.limit ?? 20, offset: options.offset ?? 0, status: options.status })}`,
    { signal: options.signal },
  );
}

export const createKnowledgeBase = (payload: {
  readonly name: string;
  readonly description: string | null;
  readonly chunk_size?: number;
  readonly chunk_overlap?: number;
}): Promise<KnowledgeBaseDetail> =>
  apiRequest("/ai/rag/knowledge-bases", {
    body: JSON.stringify(payload),
    method: "POST",
  });

export const getKnowledgeBase = (id: string, signal?: AbortSignal) =>
  apiRequest<KnowledgeBaseDetail>(`/ai/rag/knowledge-bases/${encodeURIComponent(id)}`, {
    signal,
  });

export const attachDatasetVersion = (
  knowledgeBaseId: string,
  datasetVersionId: string,
): Promise<KnowledgeBaseDatasetVersion> =>
  apiRequest(
    `/ai/rag/knowledge-bases/${encodeURIComponent(knowledgeBaseId)}/dataset-versions`,
    {
      body: JSON.stringify({ dataset_version_id: datasetVersionId }),
      method: "POST",
    },
  );

export const detachDatasetVersion = (
  knowledgeBaseId: string,
  datasetVersionId: string,
): Promise<void> =>
  apiRequest(
    `/ai/rag/knowledge-bases/${encodeURIComponent(knowledgeBaseId)}/dataset-versions/${encodeURIComponent(datasetVersionId)}`,
    { method: "DELETE" },
  );

export const buildKnowledgeBase = (knowledgeBaseId: string): Promise<RAGIndexBuild> =>
  apiRequest(`/ai/rag/knowledge-bases/${encodeURIComponent(knowledgeBaseId)}/build`, {
    method: "POST",
  });

export const cancelKnowledgeBaseBuild = (
  knowledgeBaseId: string,
): Promise<RAGIndexBuild> =>
  apiRequest(
    `/ai/rag/knowledge-bases/${encodeURIComponent(knowledgeBaseId)}/cancel-build`,
    { method: "POST" },
  );

export function listKnowledgeBaseBuilds(
  knowledgeBaseId: string,
  options: {
    readonly limit?: number;
    readonly offset?: number;
    readonly signal?: AbortSignal;
  } = {},
): Promise<RAGIndexBuildPage> {
  return apiRequest(
    `/ai/rag/knowledge-bases/${encodeURIComponent(knowledgeBaseId)}/builds${queryString({ limit: options.limit ?? 20, offset: options.offset ?? 0 })}`,
    { signal: options.signal },
  );
}

export const searchKnowledgeBase = (
  knowledgeBaseId: string,
  query: string,
  topK: number,
  minScore = 0.05,
): Promise<RAGSearchResponse> =>
  apiRequest(`/ai/rag/knowledge-bases/${encodeURIComponent(knowledgeBaseId)}/search`, {
    body: JSON.stringify({ min_score: minScore, query, top_k: topK }),
    method: "POST",
  });

File: backend/app/schemas/rag.py
"""Strict public contracts for knowledge-base retrieval and grounded chat."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    FiniteFloat,
    StringConstraints,
    field_validator,
    model_validator,
)

from app.rag.domain import (
    GroundedOutcome,
    RAGConversationStatus,
    RAGIndexBuildStatus,
    RAGKnowledgeBaseStatus,
    RAGMessageRole,
    RAGMessageStatus,
)
from app.utils.safe_text import ensure_safe_multiline, ensure_safe_single_line

KnowledgeBaseName = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        min_length=3,
        max_length=128,
        pattern=r"^[A-Za-z0-9][A-Za-z0-9 _.-]*$",
    ),
]
OptionalDescription = Annotated[
    str, StringConstraints(strip_whitespace=True, min_length=1, max_length=2000)
]
ConversationTitle = Annotated[
    str, StringConstraints(strip_whitespace=True, min_length=1, max_length=255)
]
MessageContent = Annotated[
    str, StringConstraints(strip_whitespace=True, min_length=1, max_length=4000)
]
IdempotencyKey = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        min_length=8,
        max_length=128,
        pattern=r"^[A-Za-z0-9._:-]+$",
    ),
]


class KnowledgeBaseCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    name: KnowledgeBaseName
    description: OptionalDescription | None = None
    chunk_size: int = Field(default=1000, ge=200, le=4000)
    chunk_overlap: int = Field(default=100, ge=0, le=1000)

    @field_validator("description")
    @classmethod
    def validate_description(cls, value: str | None) -> str | None:
        return ensure_safe_multiline(value) if value is not None else None

    @model_validator(mode="after")
    def validate_chunking(self) -> KnowledgeBaseCreateRequest:
        if self.chunk_overlap >= self.chunk_size:
            raise ValueError("Chunk overlap must be smaller than chunk size.")
        return self


class DatasetVersionAttachmentRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    dataset_version_id: UUID


class DatasetVersionAttachmentResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, from_attributes=True)

    dataset_version_id: UUID
    attached_at: datetime


class KnowledgeBaseSummaryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    knowledge_base_id: UUID
    name: str
    description: str | None
    status: RAGKnowledgeBaseStatus
    embedding_provider: str
    embedding_model: str
    embedding_dimension: int = Field(gt=0)
    attached_dataset_version_count: int = Field(ge=0)
    created_at: datetime
    updated_at: datetime


class KnowledgeBaseDetailResponse(KnowledgeBaseSummaryResponse):
    chunking_configuration: dict[str, object]
    active_index_build_id: UUID | None
    indexed_document_count: int = Field(ge=0)
    indexed_chunk_count: int = Field(ge=0)
    dataset_versions: list[DatasetVersionAttachmentResponse]
    error_code: str | None
    safe_error_message: str | None
    archived_at: datetime | None


class KnowledgeBasePageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    items: list[KnowledgeBaseSummaryResponse]
    total: int = Field(ge=0)
    limit: int = Field(ge=1, le=100)
    offset: int = Field(ge=0)


class IndexBuildResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    index_build_id: UUID
    knowledge_base_id: UUID
    status: RAGIndexBuildStatus
    indexed_document_count: int = Field(ge=0)
    indexed_chunk_count: int = Field(ge=0)
    embedding_count: int = Field(ge=0)
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    cancelled_at: datetime | None
    error_code: str | None
    safe_error_message: str | None


class IndexBuildPageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    items: list[IndexBuildResponse]
    total: int = Field(ge=0)
    limit: int = Field(ge=1, le=100)
    offset: int = Field(ge=0)


class RetrievalSearchRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    query: MessageContent
    top_k: int = Field(default=5, ge=1, le=20)
    min_score: FiniteFloat = Field(default=0.05, ge=0, le=1)


class RetrievalResultResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    chunk_id: UUID
    document_id: UUID
    dataset_version_id: UUID
    rank: int = Field(ge=1, le=20)
    score: float = Field(ge=0, le=1)
    excerpt: str = Field(min_length=1, max_length=500)
    document_title: str = Field(min_length=1, max_length=255)
    page_number: int | None = Field(default=None, ge=1)
    section: str | None = Field(default=None, max_length=255)


class RetrievalSearchResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    knowledge_base_id: UUID
    results: list[RetrievalResultResponse]
    insufficient_evidence: bool


class ConversationCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    knowledge_base_id: UUID
    title: ConversationTitle | None = None

    @field_validator("title")
    @classmethod
    def validate_title(cls, value: str | None) -> str | None:
        return ensure_safe_single_line(value) if value is not None else None


class ConversationResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    conversation_id: UUID
    knowledge_base_id: UUID
    title: str
    status: RAGConversationStatus
    created_at: datetime
    updated_at: datetime
    archived_at: datetime | None


class ConversationPageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    items: list[ConversationResponse]
    total: int = Field(ge=0)
    limit: int = Field(ge=1, le=100)
    offset: int = Field(ge=0)


class MessageSubmitRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    content: MessageContent
    idempotency_key: IdempotencyKey

    @field_validator("content")
    @classmethod
    def validate_content(cls, value: str) -> str:
        return ensure_safe_multiline(value)


class MessageCitationResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    citation_id: UUID
    chunk_id: UUID
    document_id: UUID
    dataset_version_id: UUID
    rank: int = Field(ge=1, le=20)
    score: float = Field(ge=0, le=1)
    excerpt: str = Field(min_length=1, max_length=500)
    document_title: str = Field(min_length=1, max_length=255)
    page_number: int | None = Field(default=None, ge=1)
    section: str | None = Field(default=None, max_length=255)


class MessageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    message_id: UUID
    conversation_id: UUID
    reply_to_message_id: UUID | None
    role: RAGMessageRole
    content: str = Field(min_length=1, max_length=4000)
    status: RAGMessageStatus
    grounded_outcome: GroundedOutcome | None
    generation_provider: str | None
    generation_model: str | None
    citations: list[MessageCitationResponse]
    created_at: datetime
    completed_at: datetime | None
    error_code: str | None
    safe_error_message: str | None


class MessageExchangeResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    user_message: MessageResponse
    assistant_message: MessageResponse


class MessagePageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    items: list[MessageResponse]
    total: int = Field(ge=0)
    limit: int = Field(ge=1, le=100)
    offset: int = Field(ge=0)
